### PR TITLE
[fuzz] Added Least request load balancer fuzz test

### PR DESCRIPTION
**Pull request by zasweq**
Commit Message: Added Least request load balancer fuzz test.
Additional Description: 60.7% line coverage over source/common/upstream/load_balancer_impl.cc. Based on PR #13722.  21ish exec/sec on cloudtop.
Risk Level: Low
Testing: Regression Tests

### CODE REVIEW

File: test/common/upstream/BUILD
Comment (asraa) [inline test/common/upstream/BUILD:796]: what is this used for?

File: test/common/upstream/least_request_load_balancer_fuzz_test.cc
Comment (asraa) [inline test/common/upstream/least_request_load_balancer_fuzz_test.cc:30]: this feels like a special case of the code above, right?
`"0"`.

File: test/common/upstream/least_request_load_balancer_fuzz_test.cc
Comment (asraa) [inline test/common/upstream/least_request_load_balancer_fuzz_test.cc:71]: why would this happen here and not after replay?

Comment (asraa) [inline test/common/upstream/least_request_load_balancer_fuzz_test.cc:71]: (could be a follow-up, why would these be defined in a destructur?

File: test/common/upstream/least_request_load_balancer_fuzz_test.cc
Comment (asraa) [inline test/common/upstream/least_request_load_balancer_fuzz_test.cc:58]: why doesn't active requests change over time?

**Review by asraa — COMMENTED** [review]
Thanks!

File: test/common/upstream/least_request_load_balancer_fuzz_test.cc
Comment (zasweq) [reply test/common/upstream/least_request_load_balancer_fuzz_test.cc:71]: This is just to make sure that it clears host state if an exception gets thrown, meaning it won't get called after replay. This is the false negative fuzz bug that came up from load balancing yesterday.

File: test/common/upstream/least_request_load_balancer_fuzz_test.cc
Comment (zasweq) [reply test/common/upstream/least_request_load_balancer_fuzz_test.cc:58]: To document, this is what we talked about in standup. Let me brainstorm some ideas about being able to decorate action streams with more specific proto fields.

File: test/common/upstream/least_request_load_balancer_fuzz_test.cc
Comment (zasweq) [reply test/common/upstream/least_request_load_balancer_fuzz_test.cc:71]: Oh interesting point, I think that will simply things wouldn't it? Rather than clearing state explicitly after replay() action sequence and during Envoy Exception throw, on destruction. I like this idea.

Comment (zasweq) [reply test/common/upstream/least_request_load_balancer_fuzz_test.cc:71]: I moved the clear call to destructor. Since this is also a fix to #13887 I'll wait until that's merged and switch my other load balancer fuzzers over to destructor logic.

File: test/common/upstream/least_request_load_balancer_fuzz_test.cc
Comment (zasweq) [reply test/common/upstream/least_request_load_balancer_fuzz_test.cc:30]: Oh beautiful, modularity :).

File: test/common/upstream/least_request_load_balancer_fuzz_test.cc
Comment (zasweq) [reply test/common/upstream/least_request_load_balancer_fuzz_test.cc:71]: Btw, I switched the logic in PR #13887 to a destructor, which also fixed the bug. I'll leave it up to you if you want me to check in before merge.

Comment (zasweq) [reply test/common/upstream/least_request_load_balancer_fuzz_test.cc:71]: It is already there after replay.

Comment (zasweq) [reply test/common/upstream/least_request_load_balancer_fuzz_test.cc:71]: Well, until I moved it to destructor :)

File: test/common/upstream/least_request_load_balancer_fuzz_test.cc
Comment (zasweq) [reply test/common/upstream/least_request_load_balancer_fuzz_test.cc:30]: I messed around with this, I liked your idea to just call into that, but I think the ASSERT on Host Weight for each fuzz iteration is too valuable to remove. I'll leave as is.

File: test/common/upstream/least_request_load_balancer_fuzz_test.cc
Comment (zasweq) [reply test/common/upstream/least_request_load_balancer_fuzz_test.cc:58]: I think since we already have a static initialization of things currently checked into master, let's make scope of this pr to just get least request out there. I'll spend the rest of the day thinking about a design that allows decorated logic for action streams.

File: test/common/upstream/BUILD
Comment (asraa) [reply test/common/upstream/BUILD:796]: ping about this comment

File: test/common/upstream/least_request_load_balancer_fuzz_test.cc
Comment (asraa) [reply test/common/upstream/least_request_load_balancer_fuzz_test.cc:58]: Just a thought -- you could add another random byte string to `UpdateHealthFlags update_health_flags = 1;`
and more adaptly name it "UpdatePrioritySet", which updates both health flags, and if needed, active requests for that priority set.

**Review by asraa — COMMENTED** [review]
Thanks! LGTM besides one existing nit
/wait

File: test/common/upstream/BUILD
Comment (zasweq) [reply test/common/upstream/BUILD]: Sorry forgot this :) deleted.

File: test/common/upstream/least_request_load_balancer_fuzz_test.cc
Comment (zasweq) [reply test/common/upstream/least_request_load_balancer_fuzz_test.cc:58]: Yeah this is a great idea! However, I think since this pr has been sitting here for a while, let's get this merged first :). I'll keep this design decision in mind as I clean up Load Balancing Fuzzing/and or add Maglev/RingHash (I saw some weight logic in both unit tests).

**Review by asraa — APPROVED** [review]
Thank you!

**Pull request merged**
